unstructured/ingest/connector/onedrive.py

```diff
@@ -0,0 +1,155 @@
+from unstructured.ingest.interfaces import (
+    BaseConnector,
+    BaseConnectorConfig,
+    BaseIngestDoc,
+    ConnectorCleanupMixin,
+    IngestDocCleanupMixin,
+    StandardConnectorConfig,
+)
+from unstructured.ingest.logger import logger
+from unstructured.utils import requires_dependencies
+
+if TYPE_CHECKING:
+    from office365.onedrive.driveitems.driveItem import DriveItem
+
+
+MAX_MB_SIZE = 512_000_000
+
+
+@dataclass
+class SimpleOneDriveConfig(BaseConnectorConfig):
+    client_id: str
+    client_credential: str = field(repr=False)
+    user_pname: str
+    tenant: str = field(repr=False)
+    authority_url: str = field(repr=False)
+    recursive: bool = False
+
+    def __post_init__(self):
+        if not (self.client_id and self.client_credential and self.user_pname):
+            raise ValueError(
+                "Please provide one of the following mandatory values:"
+                "\n-client_id\n-client_credential\n-user_principal_name",
+            )
+        self.token_factory = self._acquire_token
+
+    @requires_dependencies(["msal"])
+    def _acquire_token(self):
+        from msal import ConfidentialClientApplication
+
+        try:
+            app = ConfidentialClientApplication(
+                authority=f"{self.authority_url}/{self.tenant}",
+                client_id=self.client_id,
+                client_credential=self.client_credential,
+            )
+            token = app.acquire_token_for_client(scopes=["https://graph.microsoft.com/.default"])
+        except ValueError as exc:
+            logger.error("Couldn't set up credentials for OneDrive")
+            raise exc
+        return token
+
+
+@dataclass
+class OneDriveIngestDoc(IngestDocCleanupMixin, BaseIngestDoc):
+    config: SimpleOneDriveConfig
+    file: "DriveItem"
+
+    def __post_init__(self):
+        self.ext = "".join(Path(self.file.name).suffixes)
+        if not self.ext:
+            raise ValueError("Unsupported file without extension.")
+
+        if self.ext not in EXT_TO_FILETYPE.keys():
+            raise ValueError(
+                f"Extension not supported. "
+                f"Value MUST be one of {', '.join([k for k in EXT_TO_FILETYPE if k is not None])}.",
+            )
+        self._set_download_paths()
+
+    def _set_download_paths(self) -> None:
+        """Parses the folder structure from the source and creates the download and output paths"""
+        download_path = Path(f"{self.standard_config.download_dir}")
+        output_path = Path(f"{self.standard_config.output_dir}")
+
+        if parent_ref := self.file.get_property("parentReference", "").path.split(":")[-1]:
+            odir = parent_ref[1:] if parent_ref[0] == "/" else parent_ref
+            download_path = download_path if odir == "" else (download_path / odir).resolve()
+            output_path = output_path if odir == "" else (output_path / odir).resolve()
+
+        dname = f'{self.file.get_property("id")}-{self.file.name}'
```

Comment:
why are we prefixing the id here? is this because we're flattening the file structure? any way we could keep the file structure and ditch the need for an id added to the filename?

Reply:
Got it, I think if we preserve the file structure there should never be two colliding documents though, right? If so, safe to remove that prefix.